[16.0][IMP] email_template_qweb: copy view together with template

NL66278 opened the pull request:
Also edit the copied template inline to enhance the usability of the module. This is done with visible language selection.

This commits prevents two common errors:
- After copy and editing the qweb template from the copy, the original template is changed (as there was no copied template);
- The user edits the template, which by default shows en_US, but nothing really changes in the customer language, leading to 'weeping and gnashing of teeth'.

NL66278 committed:
[IMP] email_template_qweb: copy view together with template

Also edit the copied template inline to enhance the usability of
the module. This is done with visible language selection.

This commits prevents two common errors:
- After copy and editing the qweb template from the copy,
  the original template is changed (as there was no copied template);
- The user edits the template, which by default shows en_US, but
  nothing really changes in the customer language, leading to
  'weeping and gnashing of teeth'.

hbrunn commented on email_template_qweb/models/mail_template.py:
```diff
@@ -1,19 +1,51 @@
 class MailTemplate(models.Model):
     _inherit = "mail.template"
 
+    def _get_edit_language_selection(self):
+        """Selection for language to edit template."""
+        active_languages = self.env["res.lang"].search(
+            [("active", "=", True), ("code", "!=", "en_US")]
+        )
+        selection = [("en_US", "English (US)")] + [
+            (lang.code, lang.name) for lang in active_languages
+        ]
+        return selection
+
+    @api.depends("edit_language", "body_view_id.arch")
+    def _compute_body_view_arch(self):
+        for this in self:
+            this.body_view_arch = this.body_view_id.with_context(
+                lang=this.edit_language
+            ).arch
+
+    def _inverse_body_view_arch(self):
+        for this in self:
+            this.body_view_id.with_context(
+                lang=this.edit_language
+            ).arch = this.body_view_arch
+
```
those go below the field definitions, cf https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#37models

hbrunn commented on email_template_qweb/models/mail_template.py:
```diff
@@ -1,19 +1,51 @@
+    edit_language = fields.Selection(
+        selection="_get_edit_language_selection",
+        default="en_US",
+        help="Set the language to edit template",
```
I'd suggest to make this nonstored and default to the interface language

NL66278 replied:
I have never seen a nonstored selection field. Do you have any example?

hbrunn commented on email_template_qweb/models/mail_template.py:
```diff
@@ -50,3 +82,26 @@ def generate_email(self, res_ids, fields):
+                    default or {},
+                    body_view_id=new_view.id,
+                )
+                default["body_view_id"] = new_view.id
```
```suggestion
```

hbrunn commented on email_template_qweb/views/mail_template.xml:
```diff
@@ -9,14 +9,21 @@
                         name="body_view_arch"
                         widget="ace"
-                        attrs="{'required': [('body_type', '=', 'qweb_view')], 'invisible': [('body_view_id', '=', False)]}"
+                        context="{'lang': edit_language}"
```
do you need this given the compute and inverse functions already set it?

hbrunn reviewed (changes requested):
I'd have preferred those quite useful but very distinct features to be two commits

hbrunn commented on email_template_qweb/__manifest__.py:
```diff
@@ -1,8 +1,8 @@
 {
     "name": "QWeb for email templates",
-    "version": "16.0.1.0.0",
+    "version": "16.0.1.1.0",
```
it's usually better to have ocabot do this on merge time

NL66278 committed:
[FIX] email_template_qweb, use sudo to edit view

For QWeb email templates, the view (template) replaces the body_html
field. So if you are allowed to edit the email template, you should be
allowed to update the corresponding QWeb template.
